src/compiler/checker.ts

```diff
@@ -439,7 +439,33 @@ module ts {
                     var declaration = forEach(result.declarations, d => isBlockOrCatchScoped(d) ? d : undefined);
 
                     Debug.assert(declaration !== undefined, "Block-scoped variable declaration is undefined");
-                    if (!isDefinedBefore(declaration, errorLocation)) {
+                    
+                    // first check if usage is lexically located after the declaration
+                    var isUsedBeforeDeclaration = !isDefinedBefore(declaration, errorLocation);
+                    if (!isUsedBeforeDeclaration) {
+                        // lexical check succedded however code still can be illegal.
+                        // - block scoped variables cannot be used in its initializers
+                        //   let x = x; // illegal but usage is lexically after definition
+                        // - in ForIn/ForOf statements variable cannot be contained in expression part
+                        //   for (let x in x)
+                        //   for (let x of x)
+
+                        // climb up to the variable declaration skipping binding patterns
+                        var variableDeclaration = <VariableDeclaration>getAncestor(declaration, SyntaxKind.VariableDeclaration);
+                        var container = getEnclosingBlockScopeContainer(variableDeclaration);
+
+                        if (variableDeclaration.parent.parent.kind === SyntaxKind.VariableStatement ||
+                            variableDeclaration.parent.parent.kind === SyntaxKind.ForStatement) {
+                            // variable statement/for statement case, use site should not be inside initializer
+                            isUsedBeforeDeclaration = isChildNode(errorLocation, variableDeclaration.initializer, container);
+                        }
+                        else if (variableDeclaration.parent.parent.kind === SyntaxKind.ForOfStatement || 
+                            variableDeclaration.parent.parent.kind === SyntaxKind.ForInStatement) {
+                                // ForIn/ForOf case - use site should not be used in expression part
+                            isUsedBeforeDeclaration = isChildNode(errorLocation, (<ForInStatement>variableDeclaration.parent.parent).expression, container);
+                        }

```

Comment:
my spec reading-fu says that it should not be allowed:
>let and const declarations define variables that are scoped to the running execution context’s LexicalEnvironment. 
>The variables are created when their containing Lexical Environment is instantiated but may not be accessed in any way until the variable’s LexicalBinding is evaluated. 

and
**Runtime semantic: Evaluation**
>LexicalBinding : BindingPattern Initializer

>Let rhs be the result of evaluating Initializer.
>Let value be GetValue(rhs).
>ReturnIfAbrupt(value).
>Let env be the running execution context’s LexicalEnvironment.
>Return the result of performing BindingInitialization for BindingPattern using value and env as the arguments.


BindingInitialization step is done as a last part of evaluation of LexicalBinding, so since evaluation is not yet finished - variables cannot be accessed